rfcs/20210510-binary-size-reduction/README.md

```diff
@@ -0,0 +1,209 @@
+their applications to fit a specific disk space quota. Currently, applications
+linked statically with oneDNN exceed this quota a lot due to the library
+implementation way. Our experiments show that binary size of statically linked
+end-user application may take up to 64 MB [1]. Further analysis revealed
+several spots that contribute most:
+
+* The implementation of threading layer through templates. This is TBB specific,
+  OMP doesn't suffer from it. Takes ~14 MB. Update: was resolved in v2.4,
+  decreasing the top level of disk space to 50 MB for [1].
+* The implementation of reorders through templates of plain C code and their
+  multiple instantiations. Takes ~9 MB.
+* The implementation of GEMM autogenerated JIT-ted kernels. Takes ~14 MB.
+* The implementation of convolution kernels apart from GEMM. Takes ~8 MB.
+* Rest functionality takes ~12 MB + ~2 MB of core API.
+
+### Updated disk space chart
+Current breakdown for future v2.4 release for disk space of certain parts:
+
+* ~2 MB for core API.
+* ~12 MB for core compute functionality: convolution, deconvolution, inner
+  product and matmul.
+* ~14 MB is GEMM auto-generated kernels (would be part of compute unless
+  replaced)
+* ~11 MB are reorders (~10 MB are instances of simple reorders which could be
+  replaced with jit-ted one)
+* ~11 MB is the rest functionality.
+
+## Background
+The reason why static linking can't solve the problem itself is the way library
+API is implemented. The call to `dnnl_primitive_desc_create` (which is also used
+underneath the correspondent C++ API) relies on statically defined
+implementation lists which contain all primitives at once. That means that if
+user calls the function (which is an essential part of the API), they get the
+full library in their application. And there are not many choices how to deal
+with such problem. One is to switch off implementations in compile time (this
+RFC proposal) or using a different API which would separate things on different
+levels (at least on primitive one).
+
+## Proposal
+To control the size, the proposal is to utilize the power of build system and
+let the user specify a set of desired functionality. The idea is to have sets of
+macros, one is external and would be defined by the build system based on user
+choices. This external set would control another set - internal - which lives
+inside the library and modifies implementation lists. Each primitive would get
+several macros to switch on/off certain implementations during compilation time.
+Further optimizations of disk space will be provided by linker which excludes
+unused code inside the end-user application unless certain scenario happens [2].
+
+The proposal is to introduce three new build options which would define the list
+of available implementations for user at build time:
+* DNNL_ENABLE_PROP_KIND. This option supports two values: TRAINING (the default)
+  and INFERENCE. Second one is a subset of first one and strips backward off.
```

Comment:
Given the option name, shouldn't it be clearer to just map `DNNL_ENABLE_PROP_KIND` to actual [dnnl_prop_kind_t](https://github.com/oneapi-src/oneDNN/blob/9a753a2b62ea910962d412f49e90ac6c483de989/include/oneapi/dnnl/dnnl_types.h#L1104-L1128) values?

Reply:
Not sure I understand the proposal here. Could you clarify, please?

Reply:
So instead of accepting INFERENCE/TRAINING, it would accept FORWARD, BACKWARD_DATA, BACKWARD_WEIGHTS, ...

Also, independently of what would be the accepted values for that option, do we need to also alter the header files based on that compile option? For example, would we still expose `dnnl::prop_kind::forward_training` and `dnnl::prop_kind::backward` when build in inference mode?

Reply:
I don't see a connection between header values and build option values to be honest. Another thing is that I don't see why data and weights should be separated for backwards. Users either want training (including both backward prop kinds) or just inference (with forward only). Not sure what the rationale here for using other names...

This proposal is about tweaking internals, not about setting a different API. If we were about changing program stuff, the proposal would be different, that's why all API is left intact. Users, who control the build, should expect that calling training mode when built for inference only would result in bunch of unimplemented statuses. I don't see any evil here. Users who don't and get binaries shouldn't experience such problems since by default everything is enabled.

Reply:
> I don't see a connection between header values and build option values to be honest. 

The connection is the name of the option :-). We would have prop_kind in our header and ENABLE_PROP_KIND as a build option and both would mean different things. So if we keep the same name, they should map to the same thing (forward_inference, bakward_data, ...). An alternative would be to choose another name for the build knob, for example `DNNL_ENABLE_WORKLOAD` or something like that.

> This proposal is about tweaking internals, not about setting a different API.

I am fine with that. Could you drop a note about that in the RFC please, so that the expected behavior is clearly laid out.

Reply:
Added a note about internals and changed the name to WORKLOAD. Thanks.